File: 1584-min-cost-to-connect-all-points/1584-min-cost-to-connect-all-points.java
// not prior for now 

class Solution {
    
    public int minCostConnectPoints(int[][] points) {
        int dist = 0, pointer = 0;
        Point[] ps = initPoints(points);
        
        while (pointer < points.length) {
            Point point = ps[pointer++];
            
            dist += point.dist == Integer.MAX_VALUE ? 0 : point.dist;
            
            int nextPoint = findNextPoint(ps, point, pointer);
            
            if (nextPoint > -1)
                swap(ps, nextPoint, pointer);
        }
        
        return dist;
    }
    
    private class Point {
        int x, y, dist;
        
        public Point(int x, int y, int dist) {
            this.x = x;
            this.y = y;
            this.dist = dist;
        }
    }
    
    private Point[] initPoints(int[][] points) {
        Point[] ps = new Point[points.length];
        
        for (int i = 0; i < points.length; i++) {
            int[] p = points[i];
            
            ps[i] = new Point(p[0], p[1], Integer.MAX_VALUE);
        }
        
        return ps;
    }
    
    private int findNextPoint(Point[] ps, Point point, int start) {
        int idx = -1, min = Integer.MAX_VALUE;
        
        for (int i = start; i < ps.length; i++) {
            int m = Math.abs(point.x - ps[i].x) + Math.abs(point.y - ps[i].y);
            
            ps[i].dist = Math.min(ps[i].dist, m);
            
            if (ps[i].dist < min) {
                min = ps[i].dist;
                idx = i;
            }
        }
        
        return idx;
    }
    
    private void swap(Point[] ps, int i, int j) {
        Point tmp = ps[i];
        
        ps[i] = ps[j];
        ps[j] = tmp;
    }
}

File: 1584-min-cost-to-connect-all-points/1584-min-cost-to-connect-all-points.js
/**
 * @param {number[][]} points
 * @return {number}
 */
const minCostConnectPoints = function(points) {
    let cost = 0, prev = 0, n = points.length;
    let result = new Array(n).fill(Number.MAX_SAFE_INTEGER);
    result[0] = 0;
    for(let j=1;j<n;j++){
        let minimum = Number.MAX_SAFE_INTEGER, temp = -1;
        for(let i=1;i<n;i++){
            if(result[i] > 0){
                result[i] = Math.min(result[i],
                Math.abs(points[i][0] - points[prev][0]) +
                Math.abs(points[i][1] - points[prev][1]));
                if(result[i] < minimum) minimum = result[i], temp =i;
            }
        }
        cost += minimum, result[temp] = 0, prev = temp;
    }
    return cost;
};